XML format equivalent to native, with Reader and Writer

massifrg opened the pull request:
A new `xml` format as described in issue #10746, and in [xml.md](https://github.com/massifrg/pandoc/blob/xml/doc/xml.md).

jgm commented:
I realized that there's another, more elegant approach you could take to the quickcheck.  Instead of guarding the instances (excluding those that aren't "good"), you could normalize.

```hs
my_property = forAll (normalize <$> arbitrary) (\doc -> reader (writer doc) == doc)
```
where `normalize` has type `Pandoc -> Pandoc` and does kind of what you were doing before:  removing consecutive Space elements, collapsing consecutive Str, etc. This can be done with `walk` and a function of type `[Inline] -> [Inline]`, and in fact I *think* you can do all the hard work just by using Text.Pandoc.Builder: (untested)

```hs
normalize :: Pandoc -> Pandoc
normalize = walk fixInlines
 where
  fixInlines :: [Inline] -> [Inline]
  fixInlines = B.toList . B.fromList
```

massifrg commented:
Thanks a lot for the advice because, by excluding ill documents, I was getting this message:
```
Test suite test-pandoc: RUNNING...
pandoc tests
  XML
    p_xml_roundtrip: FAIL (0.06s)
      *** Gave up! Passed only 5 tests; 1000 discarded tests.
```
Too few tests pass the pre-selection (they don't go beyond 10).

I was trying to do the normalization myself when I saw your message.

Since the test log outputs only the auto-generated document, how can I log the normalized one?

jgm commented:
> Since the test log outputs only the auto-generated document, how can I log the normalized one?

Hm, my understanding is that it should log the normalized one.  The `(normalize <$> arbitary)` will generate a normalized document that will be passed to the function.

massifrg commented:
Here's a snippet of the current test:
```haskell
p_xml_roundtrip :: Pandoc -> Property
p_xml_roundtrip d = isValidPandoc d ==> d'' == d'
  where
    d' = normalize d
    xml = purely (writeXML def) d'
    d'' = purely (readXML def) xml

normalize :: Pandoc -> Pandoc
normalize = walk (fix' . fixInlines)
  where
    fixInlines :: [Inline] -> [Inline]
    fixInlines = B.toList . B.fromList
    fix' :: [Inline] -> [Inline]
    fix' (Space : Space : xs) = fix' $ Space : xs
    fix' (Str s1 : Str s2 : xs) = fix' $ Str (s1 <> s2) : xs
    fix' (LineBreak : SoftBreak : xs) = fix' $ LineBreak : xs
    fix' (SoftBreak : LineBreak : xs) = fix' $ LineBreak : xs
    fix' (x : xs) = x : fix' xs
    fix' [] = []
```

How do I apply that `(normalize <$> arbitary)` you are suggesting me?

jgm commented:
You'd modify your code like this (untested):
```hs
p_xml_roundtrip :: Property
p_xml_roundtrip d = forAll (normalize <$> arbitrary) (\doc -> purely (readXML def doc >>= writeXML def))
```

massifrg commented:
Thank you for the tip.

> ```haskell
> p_xml_roundtrip :: Property
> p_xml_roundtrip d = forAll (normalize <$> arbitrary)
>   (\doc -> purely (writeXML def doc >>= readXML def) == doc)
> ```

The `purely (writeXML def doc >>= readXML def)` does not work, but my limited haskell knowledge can't explain why.

Anyway this works:

```haskell
p_xml_roundtrip :: Property
p_xml_roundtrip = forAll (normalize <$> arbitrary) testdoc
  where
    testdoc d =
      if isValidPandoc d
        then purely (readXML def) (purely (writeXML def) d) == d
        else discard
```

Now I'm discarding less (only Str that contain spaces) and normalizing more.

I must normalize recursively, because a simplification makes another one possible, so the recursion ends when the normalized document is the same.

These are the simplifications I'm doing, before the xml roundtrip:
- consecutive Str: `Str "ab", Str "cd" -> Str "abcd"`
- consecutive Emph, Strong, Underline, Strikeout, SmallCaps, Superscript, Subscript are compacted
- consecutive Space are compacted to one Space
- consecutive breaks (SoftBreak, LineBreak)
- Space around breaks are removed
- Str "" are removed
- Para []. Plain [], Header _ _ [] are removed; **is this right?**

Now the test fails on tables. I must see if they need normalization or if there's something to fix in the xml reader and writer.

jgm commented on test/Tests/XML.hs:
```diff
@@ -11,74 +11,73 @@ import Text.Pandoc.Arbitrary ()
+normalize d =
+  let textual = show d
+      normalized = normalize' d
+   in if textual == show normalized
```
This comparison of `Show` instances is going to make `normalize` much slower, and it's a bit of a code smell. Why is it needed?

jgm replied:
Couldn't you just do `if d == normalized` instead of comparing the results of `show`ing them?
(And ideally `normalize'` could be adjusted until this isn't needed any more.)

massifrg replied:
I knew in FP languages that should not be necessary, but I was not sure, so I serialized the two docs as strings.

jgm commented on test/Tests/XML.hs:
```diff
@@ -11,74 +11,73 @@ import Text.Pandoc.Arbitrary ()
-    fix' (x : xs) = x : fix' xs
-    fix' [] = []
+    simplify :: [Inline] -> [Inline]
+    simplify ils = (simplifyInlines . B.toList . B.fromList) ils
```
I was wrong about `B.toList . B.fromList` - this won't do anything.

But you can try
```hs
mconcat . map B.singleton
```
This handles some of what you're doing manually:
```
ghci> (toList . mconcat . map Text.Pandoc.Builder.singleton) [Space, Str "", Str "a", Str "b", Emph [Str "a"], Emph [Str "b"]]
[Space,Str "ab",Emph [Str "a",Str "b"]]
```
Still, it doesn't handle everything, and there are performance implications, so I think it's probably best to do everything explicitly in `normalize'`.

massifrg replied:
It looks like I have 2 problems left (at least from the failed tests I saw):
- Str containing spaces: currently they make the test doc to be discarded (it's the only condition left to discard it); I will split them in normalize (maybe there's already a function in Pandoc codebase)
- tables: they are harder to normalize, because there's no `Walkable` to apply to sections, rows and cells, which would make the normalization more modular

massifrg replied:
Since I wrote `test/Tests/XML.hs`, I found only two cases requiring a change in the Reader, and none in the Writer:
- allow empty MetaList
- fix Citations' suffix detection

I'm not sure about tables: the only file I successfully roundtrip-converted is `test/tables/planets.native`; I think I'll test all the other files in the `test/tables` directory, before finding a way to normalize tables in random test docs.

jgm replied:
> because there's no Walkable to apply to sections, rows and cells, which would make the normalization more modular

Are you saying that e.g. `walk fixInlines` won't go inside table cells? If so this seems like something we should fix in pandoc-types. Or are you trying to do something different?

massifrg replied:
It does, but how do you check for an empty TableHead or an empty Row?
Can I `walk f doc` where `f :: TableHead -> TableHead` or `f :: Row -> Row`?

The failing test documents I could see were full of this corner cases, and the result of the roundtrip looked very different.

When I find a failing test, I copy-paste the contents from the log to `testNN.native.1`, then run this script:
```sh
base=$1
~/.local/bin/pandoc -s -f native -t native -o $base.native.2 $base.native.1
~/.local/bin/pandoc    -f native -t xml    -o $base.xml.3    $base.native.2
~/.local/bin/pandoc -s -f xml    -t native -o $base.native.4 $base.xml.3
meld $base.native.2 $base.native.4 &
```
(meld is a graphical diff/merge tool)
The first conversion, from native to native, is necessary to pretty print the test document and make it diff-able with the result of the roundtrip.

jgm replied:
PS. An empty TableHead is fine. Some tables don't have heads.
An empty Row, too, in principle.  (This can happen in real tables when you have rowspans.)
Why can't your XML represent these?

massifrg replied:
I should check and, in case, modify the Writer.
If an empty Row is legal in Pandoc, the XML Writer should behave accordingly.

jgm commented:
> `purely (writeXML def doc >>= readXML def)`

Right. Looks like `purely` takes a function as argument. So, this should do it:

```hs
purely (writeXML def >=> readXML def) doc
```

`>=>` is in Control.Monad.

As for the normalization:  I would say that any normalization needs a good justification.  After all, the aim is a 1-1 match with the AST.  If you had `Str` and `Space` elements in your XML representation, then I don't see why any normalization would be needed at all.

I assume you don't want to have these because they are too verbose?

Would it be possible to introduce them while still keeping the convention that XML text sections get carved into Str and Space elements (using `Text.Pandoc.Builder.text`)?

So, for example, you could have
```xml
<Para>
<Str content=""><Space>and here is some normal text
and <Str content="consecutive"><Str "strings"><Space>
</Para>
```
to handle the odd cases.  This would give you the ability to represent the AST 1-1 without any loss, but without sacrificing readability in normal cases.

In the case of other normalizations you're doing, like collapsing consecutive Emph elements, I don't see why they are helpful for your purposes (testing the XML reader/writer).  After all, you *can* represent consecutive Emph elements.  Or is the issue that your reader is using the Text.Pandoc.Builder `mappend`, which fuses these?

You could try the following `normalize` function which just mirrors what the Builder `mappend` will do:
(untested)

```hs
normalize :: Pandoc -> Pandoc
normalize = walk fixInlines
 where
   fixInlines = B.toList . mconcat . map B.singleton
```

massifrg commented:
> > `purely (writeXML def doc >>= readXML def)`
> 
> Right. Looks like `purely` takes a function as argument. So, this should do it:
> 
> ```haskell
> purely (writeXML def >=> readXML def) doc
> ```
>
> `>=>` is in Control.Monad.

Thanks, I'll implement that.
 
> As for the normalization: I would say that any normalization needs a good justification. After all, the aim is a 1-1 match with the AST. If you had `Str` and `Space` elements in your XML representation, then I don't see why any normalization would be needed at all.
> 
> I assume you don't want to have these because they are too verbose?
> 
> Would it be possible to introduce them while still keeping the convention that XML text sections get carved into Str and Space elements (using `Text.Pandoc.Builder.text`)?

That could be a writer option, that would be also used while testing, so no normalization needed.

> In the case of other normalizations you're doing, like collapsing consecutive Emph elements, I don't see why they are helpful for your purposes (testing the XML reader/writer). After all, you _can_ represent consecutive Emph elements. Or is the issue that your reader is using the Text.Pandoc.Builder `mappend`, which fuses these?

I suspect the native reader does some normalization, since I can't find it in the reader.
For example, spaces around `LineBreak` and `SoftBreak` are kept in the xml, but then they are suppressed in the native format.

jgm commented:
> I suspect the native reader does some normalization, since I can't find it in the reader.
> For example, spaces around LineBreak and SoftBreak are kept in the xml, but then they are suppressed in the native format.

As I suggested, if you are using Builder to create the AST, it does some normalization in appending.

One option would be to stop using Builder and have the parser return `Inline` elements directly.
Then you'd avoid this normalization.

jgm commented:
> That could be a writer option, that would be also used while testing, so no normalization needed.

Maybe it doesn't need to be an option. After all, the extra elements would only be present in odd corner cases, not in regular documents.  Why not leave it enabled always?

jgm commented:
> One option would be to stop using Builder and have the parser return Inline elements directly. Then you'd avoid this normalization.

Or: keep using the Builder functions, but instead of using `mconcat` to go from a list of Inlines to an Inlines, use something like

```hs
concatMany :: [Many a] -> Many a
concatMany = Many . mconcat . map unMany
```

(Remember that `Inlines = Many Inline` and `Blocks = Many Block`, so this will work for both.)  This should avoid all the normalization built into mappend instances for Inlines and Blocks.

massifrg commented:
> Or: keep using the Builder functions, but instead of using `mconcat` to go from a list of Inlines to an Inlines, use something like
> 
> ```haskell
> concatMany :: [Many a] -> Many a
> concatMany = Many . mconcat . map unMany
> ```
> 
This made things much more predictable, and now tests are passing.
I'm still using normalization, but I could restrict it to `Str` and `Space` Inlines, and to *empty* `Para`, `Plain`, `Header` and list items blocks.

In the xml reader, I'm now using the `Table` constructor and not the builder, so tables with an inconsistent number of rows/cells pass through the roundtrip conversion unaltered.

I'm going to play with `<Space />` and `<Str content="..." />` XML elements, as you suggested, but just as a way to make normalization unnecessary.
"Well-done" documents should not need them, because their AST should not contain empty or consecutive `Str`s, or `Str` with spaces; or initial, final or consecutive `Space`s.

I don't like those elements, because they make the XML less readable and make life a little harder, when you think about working on XML documents with XML tools.
But I find them a good way to achieve consistency with native and JSON formats, without any caveat.

jgm commented:
> to empty Para, Plain, Header and list items blocks.

Empty list items and headers are possible in pandoc's Markdown, and empty Para and Plain can be produced by some other formats (e.g. `<p></p>` in HTML).
Why aren't they representable in your XML?

massifrg commented:
It's not a deliberate choice. I checked that the xml writer does represent  <Header />, <Para /> and <item /> in lists. So it's the reader to be fixed.

I wanted to get to a point  where the roundtrip test passes. Starting from that, I can now fix the remaining inconsistencies.
For some corner cases, I don't know if they are legitimate in Pandoc, so it's good to know from you.

For empty list items and headers I checked it's actually as you said.

Empty Para (and Plain), instead, are discarded by `html` and `markdown` writers.
But they are kept by `native` and `json`.
If I want to keep native/json equivalence, I should keep them in `xml` too.

massifrg commented:
Empty Header, Para, Plain and list items are now fixed in the xml reader.
The normalization in the xml test is now only about `Space` and `Str`.

massifrg commented:
I'm going to encode `Space` and `Str` xml elements this way:
- `<Space />` is a single space, multiple spaces can be specified as `<Space count="..." />`, so `<Space />` is equivalent to `<Space count="1" />`
- `<Str content="blah blah" />`, where `<Str />` is equivalent to `<Str content="" />`

`Str` and `Space` usually won't be coded as elements, just as normal text with " " spaces.
The xml elements will be useful when you have these situations in the AST:
- empty strings `Str ""`
- consecutive `Str`:  `[ ..., Str "foo", Str "bar", ... ] `
- `Str` containing spaces: `Str "foo bar"`
- consecutive spaces `[ ..., Space, Space, ... ]`
- (spaces at the start or at the end of a block; this may be useful when the writer will have indentation, to discriminate between indentation and deliberate space at the start or at the end of a paragraph; but let's leave it there, for now)

massifrg commented:
> I'm going to encode `Space` and `Str` xml elements this way:
> 
>     * `<Space />` is a single space, multiple spaces can be specified as `<Space count="..." />`, so `<Space />` is equivalent to `<Space count="1" />`
> 
>     * `<Str content="blah blah" />`, where `<Str />` is equivalent to `<Str content="" />`
> 
This is done.
As a result, the test needs no more normalization:
```haskell
module Tests.XML (tests) where

import Control.Monad ((>=>))
import Test.Tasty (TestTree)
import Test.Tasty.QuickCheck
import Tests.Helpers
import Text.Pandoc
import Text.Pandoc.Arbitrary ()

p_xml_roundtrip :: Pandoc -> Bool
p_xml_roundtrip d = d == purely (writeXML def >=> readXML def) d

tests :: [TestTree]
tests = [testProperty "p_xml_roundtrip" p_xml_roundtrip]
```

The xml format looks 1:1 equivalent to `native` and `json` formats.

massifrg commented:
I learnt how to make a benchmark, with
```sh
cabal bench
```
or
```sh
cabal bench --benchmark-options="-p xml"
```
The writer benchmarks are encouraging, because `xml` is faster than any other writer but `json`:
```
    json: OK
      509  μs ±  24 μs
    xml: OK
      870  μs ±  69 μs
    native: OK
      16.4 ms ± 1.2 ms
```
Unfortunately, for the `xml` format I get only the writer benchmark, not the reader one:
```sh
$ cabal bench --benchmark-options="-p xml"
Build profile: -w ghc-9.4.8 -O1
In order, the following will be built (use -v for more details):
 - pandoc-3.6.4 (bench:benchmark-pandoc) (first run)
Preprocessing benchmark 'benchmark-pandoc' for pandoc-3.6.4..
Building benchmark 'benchmark-pandoc' for pandoc-3.6.4..
Running 1 benchmarks...
Benchmark benchmark-pandoc: RUNNING...
All
  writers
    xml: OK
      881  μs ±  77 μs

All 1 tests passed (0.40s)
Benchmark benchmark-pandoc: FINISH
```
**How do I get the xml reader benchmark?**

Another problem is that the `-s` option gives an error with the xml output format:
```sh
$ pandoc -s -f markdown -t xml -
Could not find data file templates/default.xml
```
This does not happen with `json`, though it also lacks a default template.
**How can I prevent the missing template error?**

jgm commented:
I would guess that the XML representations are also considerably more compact than the JSON ones (because of the implicit representation of String, Space sequences).

massifrg commented:
> I would guess that the XML representations are also considerably more compact than the JSON ones (because of the implicit representation of String, Space sequences).

```sh
$ pandoc -f native -t json test/testsuite.native |wc
      1     128   52574
$ pandoc -f native -t xml test/testsuite.native |wc
    721    1827   22232
```

massifrg commented:
I changed some attributes of the XML format for better consistency with Pandoc types definition:
- `Image` `src` is now `url`
- `Link` `href` is now `url`
- `Cell` `rowspan` and `colspan` are now `row-span` and `col-span`

I've added a DTD and a RelaxNG definitions to validate XML docs against (e.g. with `xmllint`).
Unfortunately, the DTD can't specify an element having any attribute, as it's needed by pandoc types with an `Attr`.

I started with a DTD, which I'm more familiar with, then converted it with `trang`, and then refined it to:
- let arbitrary attributes in elements representing Pandoc types with an `Attr`
- specify some constraints on some attributes, like Header's level, OrderedList's start, ColSpec's col-width, Cell's col-span and row-span

I put the `pandoc-xml.dtd` and `pandoc-xml.rng` in the `tools` directory. Is it the right place?

What's left to include the xml format in pandoc (maybe marked as experimental or draft)?
How can I help with that?

jgm commented on doc/xml.md:
```diff
@@ -195,7 +193,7 @@ This snippet is from the `xml` version of `test/testsuite.native`:
     <Plain>lalune</Plain>
   </Caption>
-  <Plain><Image src="lalune.jpg" title="Voyage dans la Lune">lalune</Image></Plain>
+  <Plain><Image url="lalune.jpg" title="Voyage dans la Lune">lalune</Image></Plain>
```
Sometimes filenames are used here (e.g. suppose you're converting from LaTeX).  Is `url` really appropriate?

massifrg replied:
I followed the [documentation](https://hackage.haskell.org/package/pandoc-types-1.23.1/docs/Text-Pandoc-Definition.html#t:Target), to be consistent with Pandoc types and make the naming mnemonically easier.

But, of course, I'm not that sure myself about that.

When I chose `src` for images' Target and `href` for links, I was following the HTML fashion.
Since also XML can be associated to a CSS, I checked if you could make an arbitrary attribute behave like `src` in `<img>` and `href` in `<a>` HTML elements, but there's no way to achieve that through CSS, especially when your elements are `<Image>` and `<Link>`.

Another argument against `src` and `href` is that they represent the same field in the `Target` type.

I suppose you see a problem with `<Image url="lalune.jpg"/>`, because someone would expect `url="file://lalune.jpg"` from an attribute named "url".

jgm replied:
Yes, that was the issue roughly.  For what it's worth we usually use the variable `src` for this in the pandoc code. But you're right that the docs say "URL."

It should be the same for images and links, though, and `src` seems a bit funny for links.

jgm commented:
You can always rebase this draft PR so that it has the structure of commits (and messages) that you like.

massifrg commented:
> You can always rebase this draft PR so that it has the structure of commits (and messages) that you like.

I made a first rebase to shorten the history of my commits, and I hopefully resolved the resulting conflicts.

Then I tried a rebase to reword long commits, but the rebase is too deep, and I'm afraid of messing things up (in case I did not do it yet with the previous rebase).

So I just marked the PR as ready for review.

massifrg commented:
I realized that my rebase efforts produced some mess, because in my "xml" branch I periodically merged the main branch.

The process of re-basing forced me to solve conflicts on commits that came from those merges, on files that are not on the scope of the xml format I worked on.

I think the best way to solve that is now to discard this PR and make a new one with only the files I added/modified for xml format support.

massifrg commented:
> I would make a copy of this branch how it is (in case you need to go back to it), then fix the rebasing here. It shouldn't be too hard get rid of the irrelevant changes, you can use `git checkout HEAD^` while editing any commit point with unrelated stuff mixed in to clear those bits."

Thank you for your help @alerque.

I have also a "xmlformat" [branch](https://github.com/massifrg/pandoc/tree/xmlformat) on my pandoc fork, with only a few commits and just what is needed for the xml format.
It's a bit out of sync with "main", but it's usable, in case.

jgm commented:
Where do we stand with this?

massifrg commented:
Today I synced my pandoc repository and tried to merge (locally) the "xml" branch into "main" with success.

I'm afraid there's still the problem of commit messages' length, that I don't know how to solve; when I tried, I messed up things, so I'm not going to try again.

There are no tests except the roundtrip one, which is challenging, but not something similar to examples to learn from.

There's a documentation file in the `doc` directory, but no references in the main documentation files (the user guide).

I'm going to work on the indentation in the xml writer, but the xml format could be released as experimental anyway, to let people test it.

jgm commented:
You can reword messages using `git rebase -i HASH`, where HASH is the commit before the first one you want to reword.  An editor will pop up. You can replace all the `pick`s with `r` (for reword), and then it will give you an editor window for each commit where you can reword it.

massifrg commented:
That's what I did, trying to shorten the commit messages and the history too.
But then git started asking about a lot of conflicts, because then -- I guess -- the hashes changed.
Also, during the development of the "xml" branch, I sometimes merged the "main" branch into it, to stay in sync.
Then, during the rebase I happened to change some commits coming from "main", ending in a situation that was incompatible with "main".
Or, at least, this is the way I'm figuring the mess I made trying to make the rebase.

massifrg commented:
> Since you're only touching a few files, you could always check out a fresh version of main, check out your branch in a different directory, copy the things over that need copying, and commit as necessary. Then you don't need to fight with git rebase.

I finally did it. You find it [here](https://github.com/jgm/pandoc/pull/10914).

This PR could be closed.
